### ISSUE TITLE
feat: webgui for training unmatched/matched faces

### ISSUE THREAD

**Issue by erikarenhill** [jakowenko/double-take#9]
This is a pretty big feature, but I wanted to share the idea and I might be able to help out a bit as well as I'm a web developer myself, anyhow, here's the idea described on an epic level

It would be awesome to have a web interface for browsing the files stored under storage/matches to enable you to "verify" or "dismiss" a matched photo, if you verify a photo it will be added to the training database for that person and if you click dismiss it will be untrained (if we got a false positive for someone elses face).

Also it would be awesome to have all photos with detected faces below the configured CONFIDENCE to some kind of "unmatched" directory so you can easily tag the photos for training.

**Comment by jakowenko** [comment]
I love these ideas and would love to collaborate with you on them. What's the best way to contact you?

**Comment by erikarenhill** [comment]
> I love these ideas and would love to collaborate with you on them. What's the best way to contact you?

Maybe create discussion- issues here on GitHub as a starter? Especially as we're in different timezones it would be to have quite async anyway. Otherwise discord/mail?

**Comment by erikarenhill** [comment]
Do you think it would be feasible to add a web-frontend in a separate docker so it'll be optional to use it, or should it be built into the same container-image? Maybe it adds a bit too much to the complexity to use multiple containers right now.

Regarding technology for a possible frontend, what's your preference? I might be able to help out a bit to set up a react-based frontend for it if you think that's a good choice.

**Comment by erikarenhill** [comment]
Continuing the discussion from https://github.com/jakowenko/double-take/pull/14 yes I think saving all faces that doesnt match is what at least I personally would prefer

**Comment by jakowenko** [comment]
@erikarenhill I've pushed a beta build that includes the ability to pass `SAVE_UNMATCHED: 'true'` which will save matches to `/.storage/unmatched`. I'm not pruning these images yet, but it's something I plan on adding to prevent that folder from filling up.

**Comment by jakowenko** [comment]
@erikarenhill I've pushed a beta image that includes the ability to pass `SAVE_UNKNOWN: 'true'` which will save unknown faces to `/.storage/matched/unknown`. I'm not purging these images yet, but it's something I plan on adding to prevent that folder from filling up.

**Comment by jakowenko** [comment]
> Do you think it would be feasible to add a web-frontend in a separate docker so it'll be optional to use it, or should it be built into the same container-image? Maybe it adds a bit too much to the complexity to use multiple containers right now.
> 
> Regarding technology for a possible frontend, what's your preference? I might be able to help out a bit to set up a react-based frontend for it if you think that's a good choice.

I haven't worked with React, but do have experience in Vue. Any help is appreciated, but if it's in React, I probably won't be able to maintain it unless we keep it pretty simple for now.

**Comment by erikarenhill** [comment]
> @erikarenhill I've pushed a beta image that includes the ability to pass `SAVE_UNKNOWN: 'true'` which will save unknown faces to `/.storage/matched/unknown`.
> 
> I've also added a `PURGE_UNKNOWN` and `PURGE_MATCHES` option which both default to `48` hours. Every minute a cron job runs to look for images to delete.

cool, I'm running the beta-image now with `SAVE_UNKNOWN: 'true'` and we'll see how that turns out

**Comment by jakowenko** [comment]
I added a `/train/manage` route to the beta build that lists all images from the `/matches` directory. You can either delete the image or move it to a `/train/:name` folder for training.

I figured this was a pretty good solution for now as I've been using my matched images to retrain the model and it also gives you the ability to use the `unknown` images as well. They're all displayed on this page.

I'm resizing the images currently to display on this page with a default width of 500 pixels to reduce load, but this option is configurable with a query string param `?width=250`.

I just used Pug JS and some jQuery for now to throw this page together. Nothing fancy, but should help for now. 

I have a lot of images and the performance isn't the best, but it does work. I'm lazy loading the images, but probably need to also paginate the results or lazy load those too.

Let me know what you think.

**Comment by erikarenhill** [comment]
> I added a `/train/manage` route to the beta build that lists all images from the `/matches` directory. You can either delete the image or move it to a `/train/:name` folder for training.
> 
> I figured this was a pretty good solution for now as I've been using my matched images to retrain the model and it also gives you the ability to use the `unknown` images as well. They're all displayed on this page.
> 
> I'm resizing the images currently to display on this page with a default width of 500 pixels to reduce load, but this option is configurable with a query string param `?width=250`.
> 
> I just used Pug JS and some jQuery for now to throw this page together. Nothing fancy, but should help for now.
> 
> I have a lot of images and the performance isn't the best, but it does work. I'm lazy loading the images, but probably need to also paginate the results or lazy load those too.
> 
> Let me know what you think.

This is a great starter, and simple to use! Really speed things up to train images.

I found a small bug, when moving a file from matches to training it doesn't check if the directory exists first, e.g. I have deleted my files in train/* so when I try to add an image for training for person X it tries to move it to train/X/FILENAME where X does not exist so it fails.

**Comment by jakowenko** [comment]
> This is a great starter, and simple to use! Really speed things up to train images. Should add a text-input as well to allow training new faces previously not trained/listed in the dropdown
> 
> I found a small bug, when moving a file from matches to training it doesn't check if the directory exists first, e.g. I have deleted my files in train/* so when I try to add an image for training for person X it tries to move it to train/X/FILENAME where X does not exist so it fails.
> 
> Another thing that I have spent some time thinking about is what the recommended resolution of the face should be to be able to get a good training, e.g. if the face is a small portion of the image maybe it shouldn't be able to train because it would give a bad training model and false positives for other faces, at least within the family where some kids are quite similiar looking as their siblings

Good catch on the dropdown folders, I was returning the folders from the `/matches` directory instead of `/train`. I just pushed up a fix to address this.

I'd like to add an input box like you suggested for users who don't already have a folder in `/train` and possibly the option to bulk move/delete these images, so you don't have to go one by one.

Good question on the recommended resolution. Do you have a size you think that is ideal? I'm not sure where to best fit that into the process right now. I don't save the matches and unknown images currently in the database, so the only data I have about them is just the file itself. I do get a width of the bounding box of the face when recognition initially happens, maybe I could save the width to the filename for now without needing to put it into the db.

**Comment by erikarenhill** [comment]
I read a statement from one of the CompreFace developers on reddit (cant find the source now though) that the neural network at least in CompreFace use a resolution of 160x160 so the face should at least be of that size, but the higher the better up to about 640x480 as maximum if I recall. 

Can't find anything about deepstack recommendations, but maybe we can add an environment variable with a dimension (eg 160 => 160x160) for the minimum resolution of a boundingbox for the face to be able to send for training. Personally I think I'd use around 200-250px because otherwise similiar persons would be too close each other in a such low resolution as 160x160

**Comment by jakowenko** [comment]
I like the idea of the environment variable so the user can control that. I do get false positives from time to time with my sister, so using better resolution images should definitely help with that. I'm sure right now I have a lot of low resolution images which could be causing that. Good thing there's a quick way to remove all those and retrain!

I managed to get the `train/manage` page updated to show the width and height of the recognized face, sort images with newest first, and show how long it's been since they were saved. I also updated the UI slightly to allow the user to be able to select multiple images at once to train or delete. Any old images will just show N/A for the dimensions.

Beta build just finished pushing to docker hub, let me know what you think.

What cameras did you get? I'm currently using a few Wyze and Eufy ones, which are both 2K if using their app, but over RTSP I'm only get 1080. They work ok, but the quality of images sometimes they produce from Frigate is subpar.

<img width="300" alt="Screen Shot 2021-03-31 at 3 10 00 AM" src="https://user-images.githubusercontent.com/1081811/113104912-0995c480-91cf-11eb-85a0-9e941b391a3c.png">

**Comment by erikarenhill** [comment]
I got one Reolink 810 and one 820, they were about $80 each (no wifi, bue PoE support). Currently I use a cheap 1080 chinese camera and a 4M Hikvision 2.8mm dome (which gets some faces distorted because of the wide angle). The new Reolinks use 4mm lenses. It's possible to get full resolution from RTSP but the Reolinks only support H265 on the main-stream which cause some trouble with Frigate recordings not being able to playback through the web interface.

Will try the new beta!

**Comment by jakowenko** [comment]
Nothing should've changed with that. I just tested with a picture from my phone and it's saving out both people along with any unknowns it had during the process. Are the images just getting purged or not saving at all?

![Screen Shot 2021-03-31 at 11 23 50 AM](https://user-images.githubusercontent.com/1081811/113169563-b8f48a80-9213-11eb-9109-1d280c3e7032.png)

**Comment by erikarenhill** [comment]
Not sure really, didnt have the possibility to debug it meanwhile, but I had a friend of mine visiting and we talked about this project so I wanted to show him, he stood right in front of the camera but no faces were either detected or saved. But I will try and upload a snapshot manually to compreface it might be that it didnt detect any face from the snapshot somehow. He had a cap and sunglasses on

**Comment by jakowenko** [comment]
Ah yes, if CompreFace didn't find a face it won't save the image; I'm guessing that's what happened. Let me know what happens when you upload the snapshot manually.

Also, I slightly changed the saved image filename as well to include WxH in the filename. So if you're doing any automations or notifications which require the image URL, make sure to grab the filename property which is now published to the MQTT topic.

**Comment by erikarenhill** [comment]
I took a snapshot manually from the video and compreface didn't detect any faces. I even tried to change the det_prob_threshold to 0.1 with a manual CURL-request against CompreFace but still no faces.

**Comment by jakowenko** [comment]
base64 won't reliabley work since the OS has a limit on filename length. I'll have to think of some other ways, guess I could just use the db or write a json metadata file.

**Comment by erikarenhill** [comment]
I also think that the bounding-box coordinates should be stored outside of the image (and then drawn with css or something in the web page), to keep the bitmap clean from drawn graphics when sent for training, keeping the bounding box in the actual bitmap makes little sense except when debugging maybe. Adding confidence outside is a good idea, then you don't have to zoom into the image to see the drawn graphics that can be quite small in a high-res image.

I replaced my 4M camera yesterday for one of the new Reolinks 4k/8M. Will see how that turns out, but at least I get around 200*200px faces now at some distance.

I can e-mail you some sample images if you want to see how it works/looks like in another setup than your own, prefer not to add them here in public, just drop me your e-mail in that case.

**Comment by jakowenko** [comment]
I've updated the manage UI to include more info about the images. I ended up using the image exif data to store a JSON string, so I could easily read it later. I'm pretty happy with how it came out and it's flexible for future metadata.

It does take a longer time to read a lot of the images now with exif, so I added a default of 50 images and you can adjust it with a query string parameter, `?limit=100`.

Agree with the bounding-box idea. Though I do use Node Red to send a notification with the bounding box, so there are some use cases where it makes sense. Maybe I could add a new route that serves the images with the bounding box by including a query string param? Like `http://localhost:3000/storage/matches/:name/image.jpg?bbox=1`, similar to what Frigate does.

Do you have a link to the Reolinks you bought? Curious to see how the results are for a better camera as well.

My email is my github username @gmail.com

**Comment by jakowenko** [comment]
I moved the `storage/matches/:name/:filename` route to a controller now so you could pass `?bbox=true` to draw the bounding box. It's not saved by default now.

**Comment by jakowenko** [comment]
@erikarenhill curious if you've had a chance to check any of these changes out? I also rewrote the entire UI and moved it to Vue.JS. It's currently on the beta build and is available at the root of the API `http://YOUR-IP:PORT/`. Should be easier to expand to add more features to the frontend now.

**Comment by erikarenhill** [comment]
@jakowenko sorry for late reply here, I was away for a couple of days, will try the new beta now!

**Comment by erikarenhill** [comment]
I love the new frontend, great work! It has really been a fast paced development from nothing to a really useful and nice ui. I'll send you some samples from the cameras to your e-mail.

**Comment by jakowenko** [comment]
Thank you! Glad you like it. I added the ability to add new folders from it as well, not sure if you saw that.

Hoping to work on the filtering of names / face sizes next!

**Comment by jakowenko** [comment]
Now that the UI is a lot easier to work off, are there any other things you can think of that would be useful?

I was thinking about adding a page with a live camera view for training. So you could stand in front of it and see yourself before starting the training. Could just piggy back of the Frigate latest.jpg for now and short poll that.

Having a page to delete trained images would be good too. Currently I tag each image with just the name of the user and not the filename, so right now it's only possible to untrain all of the images at once. I'd have to make a small update to include specific file untraining.

**Comment by deputyd0ng** [comment]
just piggy-backing here regarding the web-ui.

Whats the best way to handle incorrect matches?  For example, what if it thinks the mailman is me, I dont want to train the mailmans face, but I also want to tell it that the mail man is not the face it came up with a match for?

**Comment by jakowenko** [comment]
@deputyd0ng, that's a good question. One idea I had was to create a new "unknown" name and train all incorrect matches to that name, similar to how you'd train images for anyone else. Did you have any other ideas in mind?

**Comment by deputyd0ng** [comment]
funny thing you say that, I just got home and knew it would have picked up some new faces, and there were a few from strangers I dont want to train that it thought were me or my tenant, I actually considered making a new name named "unknown" but I dont know enough about how AI works and didnt want to screw up its learning.

To be fair to the algo, the camera is mounted high and to the left of the front door, Im still waiting for my amcrest doorbell  to come in to get a proper straight-on angle for faces
Im using deepstack, I dont know how other servers work, but is there a way to tell it its wrong so it can re-evaluate its confidence levels going forward?  I dont think I would have as many inaccurate matches if I had a front-facing doorbell style cam.

I had a Nest doorbell that crapped out and is currently being RMA'd (One new england winter killed it)  and then Im going to ebay it.

If I recall correctly, if it saw a face that was below a certain threshhold it would put it aside and ask for confirmation before matching, EX: "Is this John Smith" saying yes would train it, saying no would ignore it or allow you to choose a different name to train. 

 I dont mean to throw a bunch of feature requests at you when I have zero programming skills to contribute, but ideally I would like to replicate my nest hello features, minus google, so Im thinking about my experience with their product.

Your documentation for setup+config is excellent and Im using the web interface to train faces, but, being someone who generally follows online tutorials for HA, Im having trouble using the double-take data in a meaningful way in home assistant.  Other than what I described above, the most meaningful thing I think that can be produced is a few example automations, in order to make it a full replacement of my nest, I want to use TTS to announce if a matched face is discovered in a zone I masked in frigate (front porch).  I am working on it in node-red but its a hacky-ass experience for me because Im not great with node-red.  

However, between your software and frigate, I have shut down my win10 VM in unraid and am no longer using blue iris and my coral PCIE will be here Monday.  Just wish I discovered these platforms before purchasing a BI license.

Im sorry to throw a ton of rambling info at you, but I think frigate+your product has the potential to be a major game-changer in the HA environment and change how people approach security camera setup as a whole.

I also think an unraid template is a good idea to get a bunch of new users involved, who may be able to contribute more significantly and tangibly than I can.

**Comment by jakowenko** [comment]
Hey @deputyd0ng, thanks for all the great ideas! Right now any match that falls below the `CONFIDENCE` is considered a miss, which is probably similar to what your Nest did. I also have a Nest doorbell, no issues here though from the Michigan winters 🤞 .

The way all of the detectors work is you just give it images and a label for each. To my knowledge there isn't a way to tell it that the image was wrong, other than to train it with more images, or classify the wrong images with a new name, like `unknown` or `notfound`.

Currently the frontend displays a collection of all the matched and missed images. The size of the face used for training also has an effect on the accuracy. I'm working on some updates to the UI to allow the user to filter the images better and make training the detectors easier.

<img src="https://user-images.githubusercontent.com/1081811/114824981-a3469f80-9d93-11eb-8d60-019d48a87aa3.png" />

In terms of some HA automations, I could include the zone to the published MQTT topic. You could then have your Node-Red flow publish the zone to HA to then use in your TTS service. #31 was a feature request to only fire off the detection if a person was found in the zone; sounds like it would also be useful for you too.

Would it help if matches were published to a topic with the zone name instead of the name that's found? Then you could listen for new messages on a `double-take/front-door` topic or something like?

Happy to help brainstorm with you the best solution and then work to get it added to the code base!

**Comment by erikarenhill** [comment]
One additional thought on top of the "detrain/unlearn mailman example", is that sometimes you have a snapshot with multiple faces, how does one determine which face that will be trained when you select that image?

**Comment by jakowenko** [comment]
@erikarenhill, most likely the detector will reject the image if you try to pass it one with multiple faces.

**Comment by jakowenko** [comment]
I'm going to close this issue since the production build now includes all of the UI updates. Feel free to open up a new issue if you have any other ideas for it!

**Comment by ltomes** [comment]
> @erikarenhill, most likely the detector will reject the image if you try to pass it one with multiple faces.

It appears multiple detections are matched
![image](https://github.com/jakowenko/double-take/assets/4184677/0f0d541a-48f6-41f6-b68d-76b509fbe935)

![image](https://github.com/jakowenko/double-take/assets/4184677/5f8a7135-43f1-46ef-8fd0-006ecad4f5ac)

In my example here, the detector has taken the incorrect rock, and now believes with high confidence the rock is the person.

(Source images blurred)

**Comment by prat67** [comment]
Is the manage page no longer accesible? whenever I go to http://YOUR_IP:3000/train/manage i get a blank page. 

Would love to be able to train from existing images being captured....

**Comment by MrMEEE** [comment]
Same question here..

**Comment by prat67** [comment]
So I actually just figured it out last night. 
The updated version no longer has this as a separate page to manage. 

It's actually on the main Matches page. 
You select the matches (pictures) then select a name in the top left corner from the drop down and then train

![image](https://github.com/jakowenko/double-take/assets/69695694/2cbf8e10-94b2-4b81-8470-f8675c5107e6)

**Comment by MrMEEE** [comment]
Is there any way to train a false positive?.. for some reason the wheel cap on my car often get recognized as either my wife or mother?

**Comment by Kosh42** [comment]
> Is there any way to train a false positive?.. for some reason the wheel cap on my car often get recognized as either my wife or mother?

Same here, I also get false positives on unknown faces being matched as me.
